lib/Driver/ToolChain.cpp

```diff
@@ -150,6 +151,177 @@ types::ID ToolChain::lookupTypeForExtension(StringRef Ext) const {
   return types::lookupTypeForExtension(Ext);
 }
 
+/// Return a _single_ TY_Swift InputAction, if one exists;
+/// if 0 or >1 such inputs exist, return nullptr.
+static const InputAction*
+findSingleSwiftInput(const CompileJobAction *CJA) {
+  auto Inputs = CJA->getInputs();
+  const InputAction *IA = nullptr;
+  for (auto const *I : Inputs) {
+    if (auto const *S = dyn_cast<InputAction>(I)) {
+      if (S->getType() == types::TY_Swift) {
+        if (IA == nullptr) {
+          IA = S;
+        } else {
+          // Already found one, two is too many.
+          return nullptr;
+        }
+      }
+    }
+  }
+  return IA;
+}
+
+static bool
+jobsHaveSameExecutableNames(const Job *A, const Job *B) {
+  // Jobs that get here (that are derived from CompileJobActions) should always
+  // have the same executable name -- it should always be SWIFT_EXECUTABLE_NAME
+  // -- but we check here just to be sure / fail gracefully in non-assert
+  // builds.
+  assert(strcmp(A->getExecutable(), B->getExecutable()) == 0);
+  if (strcmp(A->getExecutable(), B->getExecutable()) != 0) {
+    return false;
+  }
+  return true;
+}
+
+static bool
+jobsHaveSameOutputTypes(const Job *A, const Job *B) {
+  if (A->getOutput().getPrimaryOutputType() !=
+      B->getOutput().getPrimaryOutputType())
+    return false;
+  bool sameAdditionalOutputTypes = true;
+  types::forAllTypes([&](types::ID Type) {
+      StringRef AO = A->getOutput().getAdditionalOutputForType(Type);
+      StringRef BO = B->getOutput().getAdditionalOutputForType(Type);
+      if (AO.empty() != BO.empty())
+        sameAdditionalOutputTypes = false;
+    });
+  return sameAdditionalOutputTypes;
+}
+
+static bool
+jobsHaveSameEnvironment(const Job *A, const Job *B) {
+  auto AEnv = A->getExtraEnvironment();
+  auto BEnv = B->getExtraEnvironment();
+  if (AEnv.size() != BEnv.size())
+    return false;
+  for (size_t i = 0; i < AEnv.size(); ++i) {
+    if (strcmp(AEnv[i].first, BEnv[i].first) != 0)
+      return false;
+    if (strcmp(AEnv[i].second, BEnv[i].second) != 0)
+      return false;
+  }
+  return true;
+}
+
+bool
+ToolChain::jobIsBatchable(const Compilation &C, const Job *A) const {
+  // FIXME: There might be a tighter criterion to use here?
+  if (C.getOutputInfo().CompilerMode != OutputInfo::Mode::StandardCompile)
+    return false;
+  auto const *CJActA = dyn_cast<const CompileJobAction>(&A->getSource());
+  if (!CJActA)
+    return false;
+  return findSingleSwiftInput(CJActA) != nullptr;
+}
+
+bool
+ToolChain::jobsAreBatchCombinable(const Compilation &C,
+                                  const Job *A, const Job *B) const {
+  return (jobIsBatchable(C, A) &&
+          jobIsBatchable(C, B) &&
```

Comment:
To avoid redundant checking, you could assert these conditions instead.

Reply:
Sure, fixed.